src/msg/node.h

```diff
@@ -42,14 +55,17 @@ class CRNode {
 
     std::string GetName() const { return name_; }
 
-    bool AddJobToRunner(job_t&& job, strand_ptr_t strand);
+    bool AddJobToRunner(auto&& job, JobRunnerStrandPtr strand);
```

Comment:
IIRC this is identical to `template<class T>bool Foo(T&& job);`, but shorter.

Reply:
https://en.cppreference.com/w/cpp/language/function_template#Abbreviated_function_template

Reply:
Not identical, you won't be able to do any processing on T later on. And if on interface it forces downstream to use C++20 as well, which makes it even worse.

Reply:
> Not identical, you won't be able to do any processing on T later on. 

You still can do by using `decltype`. Actually we did it in this PR for forwarding.

> And if on interface it forces downstream to use C++20 as well, which makes it even worse.

We have already done this.

Reply:
Just remind you of what C++20 features that we are currently using in this repo.

- concepts
- likely/unlikely attributes
- designated initializer
- feature testing macros

and maybe there are more

Reply:
> You still can do by using decltype

No, decltype can only be apply to the entire var, not just the T part derived.

> We have already done this.

No, I'm talking about interface, e.g. those decl in header.
auto in func signature is particularly bad because it affects the ABI, while the rest you listed is optional.

Reply:
I like new language features as much as you do, but I'm also very conservative on the application to large project, since I've seen so many compatibility issues.

Reply:
> No, decltype can only be apply to the entire var, not just the T part derived.

There are not much difference. The only one is `T&&` vs `T`.

> No, I'm talking about interface, e.g. those decl in header.
> auto in func signature is particularly bad because it affects the ABI, while the rest you listed is optional. 

Concepts are in headers.

Also, we enabled C++20 in our CI, and never tested (or want to support) it with other standards. I think the discussion of supporting pre-C++20 downstream went far outscoped.

Reply:
> There are not much difference.

Perfect forwarding will be different.

> Concepts are in headers.

Concept can actually be ignored quite easily with macro.

> Also, we enabled C++20 in our CI, and never tested (or want to support) it with other standards.

Test coverage is not a valid justification for design.

> I think the discussion of supporting pre-C++20 downstream went far outscoped.

Not that far away, for example, sometimes if you interact with python through cython or pybind you get less of choice on toolchain.

Anyway, this is just methodology from experience of forward compatibility, not to address a particular issue, so I won't be able to answer questions like "whether A is required for B" or "whether B is far from our needs".
If you insist, I won't block you as it'll take less time than this discussion to fix when needed, the concern is more on when allowing feature like this in repo, whether other colleagues can also use it properly, since there's always discussions around the abuse of auto.

Reply:
> Perfect forwarding will be different.

Actually PF is a good counter example. `forward<T>(x)` and `forward<decltype(x)>(x)` gives you the same thing.

> Not that far away, for example, sometimes if you interact with python through cython or pybind you get less of choice on toolchain.

Honestly I don't know what it looks like, but if that's the problem, we will need to set up a new CI build for that. Otherwise everything is just "I think xxx will/won't work".

> Anyway, this is just methodology from experience of forward compatibility, not to address a particular issue, so I won't be able to answer questions like "whether A is required for B" or "whether B is far from our needs".
> If you insist, I won't block you as it'll take less time than this discussion to fix when needed, the concern is more on when allowing feature like this in repo, whether other colleagues can also use it properly, since there's always discussions around the abuse of auto.

For the new language/standard library features, I do believe that we needs some sort of rules. For now I think the rule is, if it passes the CI, and all mainstream tool chains (gcc, llvm and msvc) fully support it, then it is ok to use. If you think the rule should be more detailed, you may add a readme so that everyone can follow.

For this particular one, I have no problem of ditching auto here, but I remember we also have it somewhere else in our codebase, so if you have some opinions, we may need some documented rules.